### PR TITLE
Adding the PATCH method to the REST API (webservices) + API cache management FIX

### PR DESCRIPTION
**Pull request by nsorosac**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add a PATCH method to PrestaShop's REST API
| Type?             | improvement
| Category?         | WS
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | check the description below
| How to test?      | Using Postman or Insomnia, you can try to PATCH a resource instead of updating with PUT. You won't need to give a full XML of the resource, as long as you give the "id" and the other fields you want to update.
| Possible impacts? | The REST API (especially the "specialmanagement" parts of the API that shouldn't be concerned by this PR)

> Please note that the structure of database need to be updated because of the PREFIX_webservice_permission.method ENUM values.

---
This PR addresses 2 things:
1. An improvement on the API (not a bug)
2. A fix to make cache handling work in the API, but this feature is not documented anywhere, so it's probably not used by anyone... I stumbled upon this feature when I was reverse-engineering the webservice :-)

If you try to **PUT** a resource without all required fields, you get a **400 Bad request** status:
image

With this PR, you can make the same request with the **PATCH** method, you'll end up with a **200 Ok** status:
image

To sum up:
- **PUT** method is not modified: you still use **PUT** to replace an entire resource (all of its fields)
- **PATCH** method is added: you can use **PATCH** to partially update a resource (part of its fields)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27952)

### CODE REVIEW

**Comment by nsorosac** [comment]
Please note that the structure of database need to be updated because of the `PREFIX_webservice_permission.method` ENUM values.

**Review by Progi1984 — DISMISSED** [review]
@nsorosac Thank you for your contribution. :hugs: 

Could you explain in [an issue your initial bug for this PR](https://github.com/PrestaShop/PrestaShop/issues/new/choose) (or may be link to an issue) ?

Thanks :+1:

**Comment by nsorosac** [comment]
> @nsorosac Thank you for your contribution. hugs
> 
> Could you explain in [an issue your initial bug for this PR](https://github.com/PrestaShop/PrestaShop/issues/new/choose) (or may be link to an issue) ?
> 
> Thanks +1

@Progi1984 there is no bug related since this PR addresses 2 things:

1. An improvement on the API (not a bug)
2. A fix to make cache handling work in the API, but this feature is not documented anywhere, so it's probably not used by anyone... I stumbled upon this feature when I was reverse-engineering the webservice :-)

**Review by saulaski — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by matks — COMMENTED** [review]
Hi @nsorosac this seems like a nice contribution but your PR will be tested by QA team and they need a clear list of steps to be able to observe what you did and explore it to make sure it has no issues.

> Using Postman or Insomnia, you can try to PATCH a resource instead of updating with PUT. You won't need to give a full XML of the resource, as long as you give the "id" and the other fields you want to update.

Is quite broad, could you make it more detailed 😉 ?

<br class="Apple-interchange-newline">

**Comment by nsorosac** [comment]
@matks if you try to PUT a resource without all required fields, you get a **400 Bad request** status:
image

With this PR, you can make the same request with the PATCH method, you'll end up with a **200 Ok** status:
image

To sum up:
- PUT method is not modified: you still use PUT to replace an entire resource (all of its fields)
- PATCH method is added: you can use PATCH to partially update a resource (part of its fields)

BTW, this is a very common way to behave for a REST API :-)

**Review by MeKeyCool — DISMISSED** [review]

File: classes/webservice/WebserviceRequest.php [hunk 1]
```diff
@@ -1447,6 +1453,16 @@ public function executeEntityPut()
         return $this->saveEntityFromXml(200);
     }
 
+    /**
+     * Execute PATCH method on a PrestaShop entity.
+     *
+     * @return bool
+     */
+    public function executeEntityPatch()
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
    public function executeEntityPatch(): bool
```

File: install-dev/data/db_structure.sql [hunk 2]
```diff
@@ -2139,7 +2139,7 @@ CREATE TABLE `PREFIX_webservice_permission` (
   `id_webservice_permission` int(11) NOT NULL AUTO_INCREMENT,
   `resource` varchar(50) NOT NULL,
   `method` enum(
-    'GET', 'POST', 'PUT', 'DELETE', 'HEAD'
+    'GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD'
```
Comment (Progi1984) [inline install-dev/data/db_structure.sql:2142]: Could you create a PR on `autoupgrade` for the migration ? (Not blocking this PR)

**Review by Progi1984 — COMMENTED** [review]
@nsorosac Thank you for your contribution. I have brought you two comments.

File: install-dev/data/db_structure.sql (hunk 2 above)
Comment (nsorosac) [reply install-dev/data/db_structure.sql:2142]: https://github.com/PrestaShop/autoupgrade/pull/479

Comment (Progi1984) [reply install-dev/data/db_structure.sql:2142]: Fixed.

File: classes/webservice/WebserviceRequest.php (hunk 1 above)
Comment (Progi1984) [reply classes/webservice/WebserviceRequest.php]: Ping @nsorosac

**Review by Progi1984 — COMMENTED** [review]
@nsorosac A last item and it's approval for me.

**Review by atomiix — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by hibatallahAouadni — APPROVED** [review]
Hello @nsorosac 

I tested all these methods:

1. **GET:** :heavy_check_mark: 
2. **POST:** :heavy_check_mark: 
3. **PUT:** :heavy_check_mark: 
4. **DELETE:** :heavy_check_mark: 
5. **HEAD:** :heavy_check_mark: 
6. **_PATCH:_** :heavy_check_mark: it changed the carrier even without adding the required parameters
image


Thanks!

**Comment by Progi1984** [comment]
Thanks @nsorosac & @hibatallahAouadni

**Comment by dkarvounaris** [comment]
Oh well, I backported this to the latest 1.7 and tested it. I needed it as on a current project we are managing about 300,000 products through API, and doing all kinds of product, stock and price updates almost constantly. 

It doesn't work and causes an internal server error, not because of any version incompatibility (actually it is fully compatible to 1.7), but because there's something missing. I guess the tests don't cover testing if making a PATCH request works?

I haven't checked out the full repo so I can't do a pull request right now, but here for documenting at least commenting the issue:

https://github.com/axome/PrestaShop/blob/1d85ccb5708f6bc22dca216b3129a9f4f87894b1/classes/webservice/WebserviceRequest.php#L1648 checks a constant `WebServiceRequest:HTTP_PATCH` which is missing.

I added:
```
    const HTTP_PATCH = 8;
```
at the top of the class and requests are now going through without error.

Btw, looks like the Prestashop Webservice Library also still misses handling PATCH (which I coded myself - I'm trying to submit it as pull request next week, right now my schedule is too tight, I apologize). Just mentioning, as v8 probably shouldn't be released when parts are still missing here and there.

**Comment by Seb2nim** [comment]
Hi @dkarvounaris,
Did you success to create a working patch for 1.7.8 ?
Would you be able to share ?
Regards, Seb

**Comment by dkarvounaris** [comment]
@Seb2nim Yes, and we have it for several weeks in production now, doing daily tens of thousands stock, price and description  updates with PATCH, with no issues. I've just quickly extracted and at least adding 2 patch files from my commits for you here to use with the latest 1.7.8 version (might be working with any patch version actually). Unfortunately, I still don't have the time to provide some pull request for the 1.7.8, though these may be likely not be merged anyway now that v8 is out. 

[api-patch-support-27952.patch](https://github.com/PrestaShop/PrestaShop/files/10557311/api-patch-support-27952.patch)
[api-patch-support-27952-2.patch](https://github.com/PrestaShop/PrestaShop/files/10557313/api-patch-support-27952-2.patch)
